internal/statesync/dispatcher.go

```diff
@@ -19,156 +19,69 @@ var (
 	errUnsolicitedResponse = errors.New("unsolicited light block response")
 	errNoResponse          = errors.New("peer failed to respond within timeout")
 	errPeerAlreadyBusy     = errors.New("peer is already processing a request")
-	errDisconnected        = errors.New("dispatcher has been disconnected")
+	errDisconnected        = errors.New("provider has been disconnected")
 )
 
-// dispatcher keeps a list of peers and allows concurrent requests for light
-// blocks. NOTE: It is not the responsibility of the dispatcher to verify the
-// light blocks.
-type dispatcher struct {
-	availablePeers *peerlist
-	requestCh      chan<- p2p.Envelope
-	timeout        time.Duration
-
-	mtx     sync.Mutex
-	calls   map[types.NodeID]chan *types.LightBlock
+// dispatcher multiplexes concurrent requests by multiple peers for light blocks.
+// Only one request per peer can be sent at a time
+// NOTE: It is not the responsibility of the dispatcher to verify the light blocks.
+type Dispatcher struct {
+	// the channel with which to send light block requests on
+	requestCh chan<- p2p.Envelope
+	// timeout for light block delivery (immutable)
+	timeout time.Duration
+
+	mtx sync.Mutex
+	// all pending calls that have been dispatched and are awaiting an answer
+	calls map[types.NodeID]chan *types.LightBlock
+	// signals whether the underlying reactor is still running
 	running bool
 }
 
-func newDispatcher(requestCh chan<- p2p.Envelope, timeout time.Duration) *dispatcher {
-	return &dispatcher{
-		availablePeers: newPeerList(),
-		timeout:        timeout,
-		requestCh:      requestCh,
-		calls:          make(map[types.NodeID]chan *types.LightBlock),
-		running:        true,
+func NewDispatcher(requestCh chan<- p2p.Envelope, timeout time.Duration) *Dispatcher {
```

Comment:
> Regarding timeout: Can we pass a context.Context to deal with that instead? We're using time.After in a leaky way to implement this now, and a context would both avert that issue and reduce the seams in the package API where we need to worry about timeout information.

Having a timeout here does strike me as redundant so long as the reactor/block providers sets appropriate timeouts in the context. I think I carried this behavior over from light/provider/http/http.go#http which has a global timeout alongside the use of contexts (I think the introduction of contexts came afterwards). I will remove timeout within the dispatcher.

Reply:
> The shared channel here is concerning. I think at minimum we should document when/if this channel gets closed, and by whom, and if there are any constraints on whether it may (or must) be buffered. Ideally I'd love to see this replaced by a function or an interface rather than a naked channel, but that seems like a bigger change to the existing API.

FYI, these are all new to 0.35 so now is as good a time as any to suggest better patterns